File: low-level-design/design-patterns/golang/structural/facade-pattern/facade/facade.go
package facade

type Facade interface {
	Create() error
	Destroy() error
}

func NewGCPFacade() *GCPFacade {
	return &GCPFacade{
		blockStorageManager: NewGCPBlockStorageManager(),
		networkManager:      NewGCPNetworkManager(),
		vmManager:           NewGCPVMManager(),
	}
}

type GCPFacade struct {
	blockStorageManager GCPBlockStorageManager
	networkManager      GCPNetworkManager
	vmManager           GCPVMManager
}

func (f *GCPFacade) Create() error {
	if err := f.blockStorageManager.Create(); err != nil {
		return err
	}
	if err := f.networkManager.Create(); err != nil {
		return err
	}
	if err := f.vmManager.Create(); err != nil {
		return err
	}
	return nil
}

func (f *GCPFacade) Destroy() error {
	if err := f.blockStorageManager.Delete(); err != nil {
		return err
	}
	if err := f.networkManager.Delete(); err != nil {
		return err
	}
	if err := f.vmManager.Delete(); err != nil {
		return err
	}
	return nil
}

func NewAzureFacade() *AzureFacade {
	return &AzureFacade{
		blockStorageManager: NewAzureBlockStorageManager(),
		networkManager:      NewAzureNetworkManager(),
		vmManager:           NewAzureVMManager(),
	}
}

type AzureFacade struct {
	blockStorageManager AzureBlockStorageManager
	networkManager      AzureNetworkManager
	vmManager           AzureVMManager
}

func (f *AzureFacade) Create() error {
	if err := f.blockStorageManager.Create(); err != nil {
		return err
	}
	if err := f.networkManager.Create(); err != nil {
		return err
	}
	if err := f.vmManager.Create(); err != nil {
		return err
	}
	return nil
}
func (f *AzureFacade) Destroy() error {

	if err := f.blockStorageManager.Delete(); err != nil {
		return err
	}
	if err := f.networkManager.Delete(); err != nil {
		return err
	}
	if err := f.vmManager.Delete(); err != nil {
		return err
	}
	return nil
}

File: low-level-design/design-patterns/python/structural/facade-pattern/facade.py
from gcp_manager import  GCPBlockStorageManager,GCPNetworkManager,GCPVMManager
from azure_manager import AzureBlockStorageManager,AzureNetworkManager,AzureVMManager
from abc import ABC,abstractmethod 


class Facade(ABC):

    @abstractmethod
    def create():
        pass 
    
    @abstractmethod
    def destroy():
        pass


class GCPFacade(Facade):
    def __init__(self): 
        self.block_storage_manager = GCPBlockStorageManager()
        self.network_manager = GCPNetworkManager()
        self.vm_manager = GCPVMManager()

    def create(self):
        self.block_storage_manager.create()
        self.network_manager.create()
        self.vm_manager.create()
    def destroy(self):
        self.block_storage_manager.delete()
        self.network_manager.delete()
        self.vm_manager.delete  


class AzureFacade(Facade):
    def __init__(self): 
        self.block_storage_manager = AzureBlockStorageManager()
        self.network_manager = AzureNetworkManager()
        self.vm_manager = AzureVMManager()

    def create(self):
        self.block_storage_manager.create()
        self.network_manager.create()
        self.vm_manager.create()
    def destroy(self):
        self.block_storage_manager.delete()
        self.network_manager.delete()
        self.vm_manager.delete()



